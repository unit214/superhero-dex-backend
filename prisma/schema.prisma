// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Token {
  id         Int     @id @default(autoincrement())
  address    String  @unique
  symbol     String  @default("n/a")
  name       String  @default("n/a")
  decimals   Int     @default(-1)
  malformed  Boolean @default(false)
  noContract Boolean @default(false)
  pairs0     Pair[]  @relation("Pair_Token0")
  pairs1     Pair[]  @relation("Pair_Token1")
  listed     Boolean @default(false)
}

model Pair {
  id                          Int                               @id @default(autoincrement())
  address                     String                            @unique
  token0                      Token                             @relation("Pair_Token0", fields: [t0], references: [id])
  t0                          Int
  token1                      Token                             @relation("Pair_Token1", fields: [t1], references: [id])
  t1                          Int
  liquidityInfo               PairLiquidityInfo?                @relation("Pair_Info")
  synchronized                Boolean
  liquidityInfoHistory        PairLiquidityInfoHistory[]
  liquidityInfoHistoryV2      PairLiquidityInfoHistoryV2[]
  liquidityInfoHistoryError   PairLiquidityInfoHistoryError[]
  liquidityInfoHistoryV2Error PairLiquidityInfoHistoryV2Error[]
}

model PairLiquidityInfo {
  id          Int    @id
  pair        Pair   @relation("Pair_Info", fields: [id], references: [id])
  totalSupply String
  reserve0    String
  reserve1    String
  height      Int
}

model PairLiquidityInfoHistory {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  totalSupply    String
  reserve0       String
  reserve1       String
  height         Int
  microBlockHash String
  microBlockTime BigInt
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashUniqueIndex", [pairId, microBlockHash])
}

model PairLiquidityInfoHistoryError {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashErrorUniqueIndex", [pairId, microBlockHash, error])
}

model PairLiquidityInfoHistoryV2 {
  id              Int      @id @default(autoincrement())
  pair            Pair     @relation(fields: [pairId], references: [id])
  pairId          Int
  eventType       String
  height          Int
  microBlockTime  BigInt
  logIndex        Int
  microBlockHash  String
  transactionHash String
  totalSupply     String?
  reserve0        String?
  reserve1        String?
  deltaReserve0   String?
  deltaReserve1   String?
  fiatPrice       String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex])
}

model PairLiquidityInfoHistoryV2Error {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  logIndex       Int
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex, error])
}
