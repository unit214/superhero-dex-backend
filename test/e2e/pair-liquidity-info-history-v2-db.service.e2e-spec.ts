import { Test, TestingModule } from '@nestjs/testing';
import { Decimal } from '@prisma/client/runtime/library';

import { PairLiquidityInfoHistoryV2DbService } from '@/database/pair-liquidity-info-history/pair-liquidity-info-history-v2-db.service';
import { PrismaService } from '@/database/prisma.service';
import { EventType } from '@/tasks/pair-liquidity-info-history-importer/pair-liquidity-info-history-importer-v2.service';
import {
  historyEntry1,
  historyEntry2,
  historyEntry3,
  historyEntry4,
  pair1,
  pair2,
  pair3,
  token1,
  token2,
  token3,
} from '@/test/mock-data/pair-liquidity-info-history-mock-data';

describe('PairLiquidityInfoHistoryV2DbService', () => {
  let service: PairLiquidityInfoHistoryV2DbService;
  let prismaService: PrismaService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PairLiquidityInfoHistoryV2DbService, PrismaService],
    }).compile();

    service = module.get<PairLiquidityInfoHistoryV2DbService>(
      PairLiquidityInfoHistoryV2DbService,
    );
    prismaService = module.get<PrismaService>(PrismaService);
  });

  beforeEach(async () => {
    await prismaService.token.createMany({ data: [token1, token2, token3] });
    await prismaService.pair.createMany({ data: [pair1, pair2, pair3] });
    await prismaService.pairLiquidityInfoHistoryV2.createMany({
      data: [historyEntry1, historyEntry2, historyEntry3, historyEntry4],
    });
  });

  afterEach(async () => {
    await prismaService.pairLiquidityInfoHistoryV2.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  afterAll(async () => {
    await prismaService.$disconnect();
  });

  describe('upsert', () => {
    it('should correctly upsert an existing entry', async () => {
      const updatedEntry = {
        pairId: historyEntry2.pairId,
        eventType: EventType.PairBurn,
        reserve0: new Decimal(500),
        reserve1: new Decimal(500),
        deltaReserve0: new Decimal(-500),
        deltaReserve1: new Decimal(-500),
        fiatPrice: new Decimal(0),
        height: 200002,
        microBlockHash: historyEntry2.microBlockHash,
        microBlockTime: 2000000000002n,
        transactionHash: historyEntry2.transactionHash,
        transactionIndex: 200002n,
        logIndex: historyEntry2.logIndex,
      };
      await service.upsert(updatedEntry);
      const entry = await prismaService.pairLiquidityInfoHistoryV2.findFirst({
        where: { id: historyEntry2.id },
      });
      expect(entry?.reserve0).toEqual(new Decimal(500));
    });

    it('should correctly insert an new entry', async () => {
      const newEntry = {
        pairId: 1,
        eventType: EventType.PairBurn,
        reserve0: new Decimal(500),
        reserve1: new Decimal(500),
        deltaReserve0: new Decimal(-500),
        deltaReserve1: new Decimal(-500),
        fiatPrice: new Decimal(0),
        height: 500005,
        microBlockHash: 'mh_entry5',
        microBlockTime: 5000000000005n,
        transactionHash: 'th_entry5',
        transactionIndex: 500005n,
        logIndex: 1,
      };
      await service.upsert(newEntry);
      const entry = await service.getLastlySyncedLogByPairId(1);
      expect(entry?.microBlockHash).toEqual('mh_entry5');
    });
  });

  describe('getLastlySyncedLogByPairId', () => {
    it('should correctly return the last synced log for a given pairId', async () => {
      const result1 = await service.getLastlySyncedLogByPairId(1);
      const result2 = await service.getLastlySyncedLogByPairId(2);
      expect(result1?.id).toEqual(historyEntry2.id);
      expect(result2?.id).toEqual(historyEntry4.id);
    });
  });
});
