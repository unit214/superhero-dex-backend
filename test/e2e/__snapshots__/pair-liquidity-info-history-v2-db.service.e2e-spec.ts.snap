// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairLiquidityInfoHistoryV2DbService upsert should correctly upsert an existing entry 1`] = `
{
  "deltaReserve0": "-500",
  "deltaReserve1": "-500",
  "eventType": "PairBurn",
  "fiatPrice": "0",
  "height": 200002,
  "id": 222,
  "logIndex": 1,
  "microBlockHash": "mh_entry2",
  "microBlockTime": 2000000000002n,
  "pairId": 1,
  "reserve0": "500",
  "reserve1": "500",
  "transactionHash": "th_entry2",
  "transactionIndex": 200002n,
}
`;
